### ISSUE TITLE
The output of tf.layers.conv2d in the wasm backend is all 0

### ISSUE THREAD

**Issue by liliquan0118** [tensorflow/tfjs#5907]
<em>Please make sure that this is a bug. As per our
[GitHub Policy](https://github.com/tensorflow/tensorflow/blob/master/ISSUES.md),
we only address code/doc bugs, performance issues, feature requests and
build/installation issues on GitHub. tag:bug_template</em>

**System information**
- Have I written custom code (as opposed to using a stock example script provided in TensorFlow.js):
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04):Linux Ubuntu 20.04
- Mobile device (e.g. iPhone 8, Pixel 2, Samsung Galaxy) if the issue happens on mobile device:
- TensorFlow.js installed from (npm or script link): script link
- TensorFlow.js version (use command below): 3.11.0
- Browser version: chrome 95.0.4638.69 (Official Build) (64-bit)
- Tensorflow.js Converter Version:

**Describe the current behavior**
```
var weight0=[[[[0.147626, 0.0611362, 0.4833322],[0.1071634,0.3544759,0.188389],[0.4035106,-0.1818119,0.3718792]]],
      [[[-0.1369221, 0.1789438, 0.0008314], [0.2409519 , -0.0587831, 0.5853602], [-0.2225002, 0.0479231 , 0.2069922]]]];
    var weight1= [0, 0, 0];
    var weights=[tf.tensor(weight0),tf.tensor(weight1)];
    var args={"filters": 3, "kernelSize": [2, 1], "strides": [5, 2], "padding": "same", "dataFormat": "channelsLast","dilationRate": 1,"weights":weights}
    var iarr=new Array(27).fill(3);
    var input = tf.tensor(iarr,[1,3,3,3]);
    var layer=tf.layers.conv2d(args);
    var layer_prediction=layer.apply(input);
    layer_prediction.print();
```
For the parameter configuration in the above code, the output of conv2d at the wasm backend is all 0, as follows
![image](https://user-images.githubusercontent.com/68681463/144104470-325495d8-e29f-4f3a-bc6b-1eecd54e3951.png)
However, for the same parameters and inputs, the output of conv2d in the webgl backend is different.
![image](https://user-images.githubusercontent.com/68681463/144104697-5d004bbe-bc25-43b1-b978-6b2fb3f49e0e.png)

`When I change strides to [2,2], the result of the webgl backend and the result of wasm become the same. The result is as follows`
![image](https://user-images.githubusercontent.com/68681463/144105748-1e36add3-c062-4677-8132-c7f80b496848.png)

**Describe the expected behavior**
For the same input and parameters, conv2d in different backends should output the same result.

**Standalone code to reproduce the issue**
Provide a reproducible test case that is the bare minimum necessary to generate
the problem. If possible, please share a link to Colab/CodePen/any notebook.
```
<!DOCTYPE html>
<html lang="en">
<head>
  <meta charset="UTF-8">

</head>
<body>
<script src="https://cdn.jsdelivr.net/npm/@tensorflow/tfjs"> </script>
<script src="https://cdn.jsdelivr.net/npm/@tensorflow/tfjs-backend-webgl"></script>
<script src="https://cdn.jsdelivr.net/npm/@tensorflow/tfjs-backend-wasm/dist/tf-backend-wasm.js"></script>
<script>

  async function prediction (backend){
    await tf.setBackend(backend);
    await tf.ready();
    var weight0=[[[[0.147626, 0.0611362, 0.4833322],[0.1071634,0.3544759, 0.188389 ],[0.4035106,-0.1818119, 0.3718792]]],
      [[[-0.1369221, 0.1789438 , 0.0008314], [0.2409519 , -0.0587831, 0.5853602], [-0.2225002, 0.0479231 , 0.2069922]]]];
    var weight1= [0, 0, 0];
    var weights=[tf.tensor(weight0),tf.tensor(weight1)];
    var args={"filters": 3, "kernelSize": [2, 1],
      "strides": [5, 2],
      "padding": "same",
      "dataFormat": "channelsLast",
      "dilationRate": 1,
      "weights":weights}
    var iarr=new Array(27).fill(3);
    var input = tf.tensor(iarr,[1,3,3,3]);
    var layer=tf.layers.conv2d(args);
    var layer_prediction=layer.apply(input);
    layer_prediction.print();

  }
  async function run(){
    await prediction("webgl");
    await prediction("wasm");
  }
  run();



</script>
</body>
</html>

```
**Other info / logs** Include any logs or source code that would be helpful to
diagnose the problem. If including tracebacks, please include the full
traceback. Large logs and files should be attached.

**Comment by liliquan0118** [comment]
> It looks like XNN throws an error for trivial inputs like this one since you can avoid the conv. Will need to add an exception for this case in our code

thank you for your reply. I am wondering if the current result shows that the source code of tf.conv2d under the wasm backend is not robust enough? In theory, if I give an input whose syntax meets the requirements, the tf.conv2d source code should not give erroneous results. In particular, other backends can give normal results, does it prove that the current implementation of the wasm backend is wrong?

**Comment by gaikwadrahul8** [comment]
[chunnienc](https://github.com/chunnienc)

Hi, @liliquan0118 

Apologize for the delayed response and I tried to replicate the same issue with latest version and I'm getting the same output as you mentioned earlier in output log and it seems like we haven't taken care of this issue if I'm not wrong, For your reference I have added screenshot below so we will have to dig more into this issue and we'll update you soon. thank you for noticing this behaviour, I really appreciate your efforts and time. Thank you!

CC :@chunnienc

Here is output of conv2d at the `wasm` backend with all zeros and for conv2d in the `webgl` backend output is different. : 

![image](https://github.com/tensorflow/tfjs/assets/115997457/f4ba169b-5903-4b0e-a339-e75915510b6e)

After changing the strides to [2,2], the result of the `webgl` backend and the result of `wasm` become the same as follows :

![image](https://github.com/tensorflow/tfjs/assets/115997457/af2aabb4-d7cf-4221-a892-8bcb3506f17a)

**Comment by avmohansai** [comment]
Hi,
Thanks for the detailed report and reproduction code!

It looks like the SAME padding logic diverges when the stride (5) is larger than the input dimension (3). In this scenario, WebGL calculates a result, but the WASM kernel returns zeros as the stride skips the data entirely.

To get consistent results ,a workaround is to manually pad your input to match the stride and use valid padding.
We appreciate you bringing this to our attention and helping us improve the library!
